### PR TITLE
Generate reports unit test

### PR DESCRIPTION
**Pull request by leonardohscastilho**
~DO NOT MERGE THIS UNTIL CATEGORIZING CACHE MECHANISM COULD GET CLEANED~

Generate reports is the script that generates the following files as output:
 - Balance with the total sum of income and expenses 
 - Incomes categorized and grouped
 - Expenses categorized and grouped

The unit test uses a predefined configuration and transaction file in the 'tests/test_data' folder. The transaction file is defined as follows:
 - 2x entertainment category transactions
 - 2x restaurant category transactions
 - 2x internal transactions
 - 2x incomes (salary and rental)

Other tests like input checks are created as well

### CODE REVIEW

File: tests/test_generate_reports.py [hunk 1]
```diff
@@ -0,0 +1,357 @@
...
+    transactions_raw = _open_transaction_test_file(transactions_test_file_path)
+    incomes, expenses = _split_transactions(transactions_raw)
+    grouped_income, grouped_income_ref = group_transactions(
+        incomes, TransactionGroupingType.ReferenceSimilarity
+    )
+    grouped_expense, grouped_expense_ref = group_transactions(
+        expenses, TransactionGroupingType.ReferenceSimilarity
+    )
+
+    clear_user_configuration_cache()
```
Comment (diegotsutsumi) [inline tests/test_generate_reports.py:311]: The whole block reproduces a real code business logic, which can be buggy, then it defeats the purpose of testing the real code.

I suggest removing this block and carry on the test with written/validated files for what should be the ideal output.

File: tests/test_generate_reports.py [hunk 2]
```diff
@@ -0,0 +1,357 @@
...
+            transaction["referenceText"],
+            categorized_transactions_ref[transaction["groupNumber"]],
+            fallback_reference=transaction["groupName"],
+        )
+        transaction["customCategory"] = key_category
+
+    return categorized_transactions
+
+
+def test_generate_reports_output(write_json_mock: Mock) -> None:
```
Comment (diegotsutsumi) [inline tests/test_generate_reports.py:296]: Let's get rid of duplicating the business logic in unit tests and create some hard coded "expected files" that should come out of `generate_reports` when given inputs `test_transaction.json` and `test_config_file.yaml`.

It's important that we validate manually that the hard coded expected files are correct, this is a one off validation to build the test.

File: tests/test_generate_reports.py [hunk 3]
```diff
@@ -0,0 +1,357 @@
...
+) -> None:
+    with patch("personal_finances.generate_reports._write_reports"):
+        runner = CliRunner()
+        result = runner.invoke(generate_reports, command_params)
+        assert result.exit_code == 0
+        cache_user_configuration_mock.assert_called_once_with(user_file_path)
+        open_mock.assert_called_once_with(transactions_file_path, "r")
+
+
+def _open_transaction_test_file(transactions_file_path: str) -> List[SimpleTransaction]:
```
[the remark this answers is not in the document]
Comment (leonardohscastilho) [reply tests/test_generate_reports.py:158]: this function has been removed

File: tests/test_generate_reports.py (hunk 1 above)
Comment (leonardohscastilho) [reply tests/test_generate_reports.py:311]: this block has been removed

File: tests/test_generate_reports.py [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+        "expense_per_group.json",
+        "expense_per_category.json",
+        "expense_categorized_transactions.json",
+    )
+
+    for idx, file_name in enumerate(expected_file_name_list):
+        assert result_calls[idx].args[0] == expected_path_name + file_name
+        assert_files_are_equal(
+            expected_file_path_prefix + file_name, result_calls[idx].args[1]
+        )
```
Comment (diegotsutsumi) [inline tests/test_generate_reports.py:186]: What about:
```
expected_result_path_prefix = (
    "reports/1970-01-01T00:00:00+00:00_2100-01-01T00:00:00+00:00/"
)

expected_content_path_prefix = "tests/test_data/expected/"

expected_file_name_list = (
    "balance.json",
    "income_per_group.json",
    "income_per_category.json",
    "income_categorized_transactions.json",
    "expense_per_group.json",
    "expense_per_category.json",
    "expense_categorized_transactions.json",
)

result_calls = {
    result_call.args[0].removeprefix(expected_result_path_prefix): {
        "file_path": result_call.args[0],
        "file_content": result_call.args[1],
    }
    for result_call in write_json_mock.call_args_list
}

for expected_file_name in expected_file_name_list:
    result_file_path = result_calls[expected_file_name]["file_path"]
    result_file_content = result_calls[expected_file_name]["file_content"]

    assert result_file_path in expected_content_path_prefix + expected_file_name
    assert_file_content_json(
        expected_content_path_prefix + expected_file_name,
        result_file_content
    )
```

It's a bit longer, but it's clearer and it drops the `idx` ordering assumption, which shouldn't matter.

**Review by diegotsutsumi — APPROVED** [review]

**Pull request merged**